tsdb/engine/tsm1/bit_reader.go

```diff
@@ -0,0 +1,136 @@
+package tsm1
+
+// NOTE(benbjohnson)
+// Copied from: github.com/dgryski/go-bitstream
+// Required copy so that the underlying reader could be reused without alloc.
+
+import "io"
+
+// A BitReader reads bits from an io.Reader.
+type BitReader struct {
+	data []byte
+
+	buf struct {
+		v uint64 // bit buffer
+		n uint   // available bits
+	}
+}
+
+// NewBitReader returns a new instance of BitReader that reads from data.
+func NewBitReader(data []byte) *BitReader {
+	b := new(BitReader)
+	b.Reset(data)
+	return b
+}
+
+// Reset sets the underlying reader on b and reinitializes.
+func (r *BitReader) Reset(data []byte) {
+	r.data = data
+	r.buf.v, r.buf.n = 0, 0
+	r.readBuf()
+}
+
+// CanReadBitFast returns true if calling ReadBitFast() is allowed.
+// Fast bit reads are allowed when at least 2 values are in the buffer.
+// This is because it is not required to refilled the buffer and the caller
+// can inline the calls.
+func (r *BitReader) CanReadBitFast() bool { return r.buf.n > 1 }
+
+// ReadBitFast is an optimized bit read.
+// IMPORTANT: Only allowed if CanReadFastBit() is true!
+func (r *BitReader) ReadBitFast() bool {
+	v := (r.buf.v&(1<<63) != 0)
+	r.buf.v <<= 1
+	r.buf.n -= 1
+	return v
+}
```

Comment:
Fast read usage can be found here: https://github.com/influxdata/influxdb/pull/6424/files#diff-4fcee93b376ee05d583de8af541239b8R181